Refactor FastNormalClosure and don't use it when kernel is abelian

ssiccha opened the pull request:
- Refactor FastNormalClosure
- Don't use FastNormalClosure if kernel is abelian

danielrademacher commented on gap/generic/kernel.gi:
```diff
@@ -77,36 +77,35 @@ end );
 # Computes randomly (it might underestimate) the normal closure of <list>
 # under conjugation by the group generated by <grpgens>.
```
Maybe add some information about <n> from the input?

ssiccha replied:
Jepp, good point.

FriedrichRober replied:
I second that. It would be nice to know what the error probability is in terms of `n` (and the maximal length of subgroup chains in `<grpgens>`).

danielrademacher commented on gap/matrix.gi:
```diff
@@ -539,8 +540,9 @@ function(ri,G)
 
   # the kernel:
+  # The kernel is abelian, so we don't need to do normal closures.
+  findgensNmeth(ri).func := FindKernelRandom;
```
As above.

danielrademacher commented on gap/matrix.gi:
```diff
@@ -239,8 +239,9 @@ function(ri, G)
           # FIXME: why don't we just compute a precise set of generators of the kernel?
           # That should be easily and efficiently possible at this point, no?
+          # The kernel is abelian, so we don't need to do normal closures.
+          findgensNmeth(ri).func := FindKernelRandom;
```
Is "findgensNmeth(ri).func" correct? I have not found it at any other point in the package. Maybe "findgensNmeth(ri).method"?

FriedrichRober replied:
Well spotted. The documentation says that `findgensNmeth(ri)`  is a record with at least two components bound, namely `method` and `args`.

ssiccha replied:
Nice. That code must have been called by the tests I think. So the docs may be outdated. I'll double check when I'm back from holidays next week.

ssiccha replied:
Oh, or it simply never got called since I put it into the wrong component. ^^

danielrademacher commented on gap/matrix.gi:
```diff
@@ -239,8 +239,9 @@ function(ri, G)
+          # The kernel is abelian, so we don't need to do normal closures.
+          findgensNmeth(ri).func := FindKernelRandom;
           findgensNmeth(ri).args[1] := 7;
-          findgensNmeth(ri).args[2] := 5;
```
I'm also not sure, what arguments of which function are stored here. Maybe some other reviewer has already worked with this?

ssiccha replied:
That should be explained in the docs of findgensNmeth. I'll check.

ssiccha replied:
These are the arguments to the function stored in `findgensNmeth(ri).method`, in this case `FindKernelRandom`.

danielrademacher commented on gap/matrix.gi:
```diff
@@ -539,8 +540,9 @@ function(ri,G)
+  # The kernel is abelian, so we don't need to do normal closures.
+  findgensNmeth(ri).func := FindKernelRandom;
   findgensNmeth(ri).args[1] := Length(ri!.blocks)+10;
-  findgensNmeth(ri).args[2] := 7;
```
As above.

danielrademacher reviewed (commented):
The function "FastNormalClosure" looks very good. I like the changes.

I also checked that the kernel is abelian and, hence, we don't need to call the normal closure function in the lines improved by @ssiccha.

fingolfin reviewed (changes requested):
This looks plausible but there are things that were pointed out in review that still need to be addressed.

ssiccha commented:
I added a comment referencing Lemma 2.3.8 from "Permutation Group Algorithms" by Akos Seress, which reads:
> **Lemma 2.3.8.** Suppose that `G = <S>` acts on H and let U ⊆ H . Let g be a
> random subproduct of S and u be a random subproduct of U . If <U > is not
> closed for the G-action then Prob(u ^ g not in <U >) ≥ 1/4.

I addressed all comments and added two small commits:
- Add small optimisation to FastNormalClosure
- Fix documentation of findgensNmeth

ssiccha commented:
Rebased onto master.

danielrademacher reviewed (approved):
I checked everything again and this looks great now. Thanks @ssiccha.

fingolfin commented on gap/generic/kernel.gi:
```diff
@@ -75,38 +75,46 @@ InstallGlobalFunction( RandomSubproduct, function(a)
+  local list2, fewGenerators, repetitions, randlist, conjugators, i, c;
+  list2 := ShallowCopy(list);
+  if Length(grpgens) = 1 then
+    return list2;
```
Note that if we instead of `gens` we passed a group object, we could test here for `HasIsAbelian(G) and IsAbelian(G)`. And in all places calling here that I found, there already was such a group object -- with the exception of a call in `IsThisSL2Natural`. But there in turn, all (?) callers to `IsThisSL2Natural` could easily be changed to pass on a group object.

One could even support both variants in this function, by turning `gens` into a local var, and calling the first argument `G`, and doing something like `if IsGroup(G) then gens := GeneratorsOfGroup(G); else gens:=G; fi;` or so

ssiccha replied:
I'll put this idea into its own issue.
